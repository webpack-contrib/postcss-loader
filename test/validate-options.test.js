/* eslint-disable global-require */
import path from "path";

import { getCompiler, compile } from "./helpers/index";

describe("validate options", () => {
  const tests = {
    execute: {
      success: [false],
      failure: [1, "test", /test/, [], {}, { foo: "bar" }],
    },
    postcssOptions: {
      success: [
        { parser: "sugarss" },
        { parser: require("sugarss") },
        { parser: require("sugarss").parse },
        { syntax: "sugarss" },
        { syntax: require("sugarss") },
        { stringifier: "sugarss" },
        { stringifier: require("sugarss") },
        { stringifier: require("sugarss").stringify },
        {
          plugins: [
            require("./fixtures/plugin/plugin")(),
            require("./fixtures/plugin/plugin"),
            ["postcss-short", { prefix: "x" }],
          ],
        },
        {
          plugins: [
            require("./fixtures/plugin/plugin")(),
            require("./fixtures/plugin/plugin"),
            { "postcss-short": { prefix: "x" } },
          ],
        },
        { plugins: { "postcss-short": { prefix: "x" } } },
        { config: true },
        { config: false },
        { config: "test/fixtures/config-scope/config/postcss.config.js" },
        {
          config: path.resolve(
            __dirname,
            "./fixtures/config-scope/config/postcss.config.js",
          ),
        },
      ],
      failure: [{ config: [] }, { config: /test/ }],
    },
    sourceMap: {
      success: [true, false],
      failure: [1, /test/, [], {}, "something"],
    },
    implementation: {
      success: [require("postcss"), "postcss"],
      failure: [1, /test/, [], {}],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === "object" && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      let compiler;

      if (
        key === "postcssOptions" &&
        // eslint-disable-next-line no-prototype-builtins
        (value.hasOwnProperty("parser") || value.hasOwnProperty("syntax"))
      ) {
        compiler = getCompiler(
          "./sss/index.js",
          {},
          {
            module: {
              rules: [
                {
                  test: /\.sss$/i,
                  use: [
                    {
                      loader: require.resolve("./helpers/testLoader"),
                      options: {},
                    },
                    {
                      loader: path.resolve(__dirname, "../src"),
                      options: { [key]: value },
                    },
                  ],
                },
              ],
            },
          },
        );
      } else {
        compiler = getCompiler("./css/index.js", { [key]: value });
      }

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === "success") {
          expect(stats.hasErrors()).toBe(false);
        } else if (type === "failure") {
          const {
            compilation: { errors },
          } = stats;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
