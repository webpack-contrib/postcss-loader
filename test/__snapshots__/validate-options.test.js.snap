// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "execute" option with "/test/" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.execute should be a boolean.
   -> Enables/Disables PostCSS parser support in 'CSS-in-JS'.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#execute"
`;

exports[`validate options should throw an error on the "execute" option with "[]" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.execute should be a boolean.
   -> Enables/Disables PostCSS parser support in 'CSS-in-JS'.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#execute"
`;

exports[`validate options should throw an error on the "execute" option with "{"foo":"bar"}" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.execute should be a boolean.
   -> Enables/Disables PostCSS parser support in 'CSS-in-JS'.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#execute"
`;

exports[`validate options should throw an error on the "execute" option with "{}" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.execute should be a boolean.
   -> Enables/Disables PostCSS parser support in 'CSS-in-JS'.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#execute"
`;

exports[`validate options should throw an error on the "execute" option with "1" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.execute should be a boolean.
   -> Enables/Disables PostCSS parser support in 'CSS-in-JS'.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#execute"
`;

exports[`validate options should throw an error on the "execute" option with "test" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.execute should be a boolean.
   -> Enables/Disables PostCSS parser support in 'CSS-in-JS'.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#execute"
`;

exports[`validate options should throw an error on the "implementation" option with "/test/" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be one of these:
   string | function
   -> The implementation of postcss to use, instead of the locally installed version
   -> Read more at https://github.com/webpack-contrib/postcss-loader#implementation
   Details:
    * options.implementation should be a string.
    * options.implementation should be an instance of function."
`;

exports[`validate options should throw an error on the "implementation" option with "[]" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be one of these:
   string | function
   -> The implementation of postcss to use, instead of the locally installed version
   -> Read more at https://github.com/webpack-contrib/postcss-loader#implementation
   Details:
    * options.implementation should be a string.
    * options.implementation should be an instance of function."
`;

exports[`validate options should throw an error on the "implementation" option with "{}" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be one of these:
   string | function
   -> The implementation of postcss to use, instead of the locally installed version
   -> Read more at https://github.com/webpack-contrib/postcss-loader#implementation
   Details:
    * options.implementation should be a string.
    * options.implementation should be an instance of function."
`;

exports[`validate options should throw an error on the "implementation" option with "1" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be one of these:
   string | function
   -> The implementation of postcss to use, instead of the locally installed version
   -> Read more at https://github.com/webpack-contrib/postcss-loader#implementation
   Details:
    * options.implementation should be a string.
    * options.implementation should be an instance of function."
`;

exports[`validate options should throw an error on the "postcssOptions" option with "{"config":[]}" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.postcssOptions should be one of these:
   object { config?, … } | function
   -> Options to pass through to \`Postcss\`.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#postcssOptions
   Details:
    * options.postcssOptions.config should be one of these:
      string | boolean
      -> Allows to specify PostCSS config path.
      -> Read more at https://github.com/webpack-contrib/postcss-loader#config
      Details:
       * options.postcssOptions.config should be a string.
         -> Allows to specify the path to the configuration file
       * options.postcssOptions.config should be a boolean.
         -> Enables/Disables autoloading config"
`;

exports[`validate options should throw an error on the "postcssOptions" option with "{"config":{}}" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.postcssOptions should be one of these:
   object { config?, … } | function
   -> Options to pass through to \`Postcss\`.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#postcssOptions
   Details:
    * options.postcssOptions.config should be one of these:
      string | boolean
      -> Allows to specify PostCSS config path.
      -> Read more at https://github.com/webpack-contrib/postcss-loader#config
      Details:
       * options.postcssOptions.config should be a string.
         -> Allows to specify the path to the configuration file
       * options.postcssOptions.config should be a boolean.
         -> Enables/Disables autoloading config"
`;

exports[`validate options should throw an error on the "sourceMap" option with "/test/" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.sourceMap should be a boolean.
   -> Enables/Disables generation of source maps.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#sourcemap"
`;

exports[`validate options should throw an error on the "sourceMap" option with "[]" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.sourceMap should be a boolean.
   -> Enables/Disables generation of source maps.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#sourcemap"
`;

exports[`validate options should throw an error on the "sourceMap" option with "{}" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.sourceMap should be a boolean.
   -> Enables/Disables generation of source maps.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#sourcemap"
`;

exports[`validate options should throw an error on the "sourceMap" option with "1" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.sourceMap should be a boolean.
   -> Enables/Disables generation of source maps.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#sourcemap"
`;

exports[`validate options should throw an error on the "sourceMap" option with "something" value 1`] = `
"Invalid options object. PostCSS Loader has been initialized using an options object that does not match the API schema.
 - options.sourceMap should be a boolean.
   -> Enables/Disables generation of source maps.
   -> Read more at https://github.com/webpack-contrib/postcss-loader#sourcemap"
`;
